mozilla/blurts-server
# Prevent imports of `server-only` in non-Next.js code

# References:

Jira: MNTOR-3641
Figma: N/A



# Description

The breach alert cronjob broke because it imported `BreachAlertEmail`, which imported `getPremiumSubscriptionUrl`, which imported the `server-only` module. That module is supposed to throw when it's being imported in client components, which it does, but it also throws in everything that's not explicitly recognised as a server component, i.e. in cronjobs.

We had a workaround for that, but I didn't use it. I did now.

I also added a lint rule that prevents imports from `server-only`. That wouldn't have helped in this specific case because [it doesn't check dynamic imports](https://eslint.org/docs/latest/rules/no-restricted-imports#rule-details), but did catch a bunch of others, and usually, this module wouldn't be imported dynamically.

# How to test

It's a bit of a pain, but [the PubSub instructions](https://github.com/mozilla/blurts-server?tab=readme-ov-file#pubsub) are accurate. If you enable the `BreachEmailRedesign` flag, you should hit the most important code paths; if you actually want the email to get sent, you'll have to modify the curl request to use a breach name that your account hasn't been notified of yet, use the first six characters of your user's `primary_sha1` as `hashPrefix`, and the rest as a `hashSuffix`.

# Checklist (Definition of Done)

- [ ] I've added a unit test to test for potential regressions of this bug. - This code needs more reworking to make proper unit tests feasible, unfortunately.
